lib/Echidna/UI.hs

```diff
@@ -148,22 +154,28 @@ ui :: ( MonadCatch m, MonadRandom m, MonadReader x m, MonadUnliftIO m
    -> [SolTest] -- ^ Tests to evaluate
    -> Maybe GenDict
    -> m Campaign
-ui v w ts d = let xfer bc = use hasLens >>= liftIO . writeBChan bc
-                  d' = fromMaybe defaultDict d
-                  getSeed = view $ hasLens . to seed . non (d' ^. defSeed) in do
+ui v w ts d = do
+  let d' = fromMaybe defaultDict d
+  let getSeed = view $ hasLens . to seed . non (d' ^. defSeed)
   bc <- liftIO $ newBChan 100
-  t    <- forkIO (void $ campaign (xfer bc) v w ts d)
+  ref <- liftIO $ newIORef defaultCampaign
+  let updateRef = use hasLens >>= liftIO . atomicWriteIORef ref
```

Comment:
Why are we tracking campaign state in both an IORef and a BChan? Can we simplify to one or the other?

Reply:
IORef is used to sample a value of campaign at a particular point in time. This value is then pushed as a UI update through a BChan. The state is rather tracked in IORef, BChan acts as an event queue for Brick UI. Using both is what allows for controlled UI updates / decoupling.